File: app/src/test/java/com/matchthree/ui/game/BufferedSwapGuardTest.kt
package com.matchthree.ui.game

import com.matchthree.game.model.Gem
import com.matchthree.game.model.GemType
import com.matchthree.game.model.Special
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/**
 * Tests the buffered-swap staleness rule (MECHANICS.md, Input lock): a swap
 * buffered during input lock goes stale exactly when a Hypercube entered or
 * left its pair between the gesture and the flush, compared by gem id.
 */
class BufferedSwapGuardTest {

    private var nextId = 0

    private fun gem(type: GemType = GemType.RED, special: Special? = null): Gem =
        Gem(nextId++, type, special)

    @Test
    fun `plain pair whose gems changed entirely is not stale`() {
        // Fast-player path: a plain buffered swap still executes after a cascade
        // even when both cells hold different gems by then.
        val submitA = gem()
        val submitB = gem()
        assertFalse(
            bufferedSwapIsStale(submitA, submitB, settledA = gem(), settledB = gem()),
        )
    }

    @Test
    fun `hypercube still in the pair at settle is not stale`() {
        val hyper = gem(GemType.BLUE, Special.HYPERCUBE)
        val plain = gem()
        assertFalse(bufferedSwapIsStale(hyper, plain, hyper, plain))
    }

    @Test
    fun `hypercube born or fallen into the pair is stale`() {
        // The bug case: the buffered gesture never aimed at this Hypercube.
        val plainA = gem()
        val plainB = gem()
        val settledHyper = gem(GemType.BLUE, Special.HYPERCUBE)
        assertTrue(bufferedSwapIsStale(plainA, plainB, settledHyper, plainB))
        assertTrue(bufferedSwapIsStale(plainA, plainB, plainA, settledHyper))
    }

    @Test
    fun `hypercube that left the pair is stale`() {
        val hyper = gem(GemType.BLUE, Special.HYPERCUBE)
        val plain = gem()
        assertTrue(
            bufferedSwapIsStale(hyper, plain, settledA = gem(), settledB = gem()),
        )
    }

    @Test
    fun `hypercube falling within the pair keeps the swap fresh`() {
        // Same gem id, other cell of the pair: still the Hypercube the player targeted.
        val hyper = gem(GemType.BLUE, Special.HYPERCUBE)
        val plain = gem()
        assertFalse(bufferedSwapIsStale(hyper, plain, plain, hyper))
    }

    @Test
    fun `a second hypercube entering the pair is stale`() {
        // The flush would fire a Hypercube+Hypercube combo the player never gestured at.
        val hyper = gem(GemType.BLUE, Special.HYPERCUBE)
        val plain = gem()
        val secondHyper = gem(GemType.GREEN, Special.HYPERCUBE)
        assertTrue(bufferedSwapIsStale(hyper, plain, hyper, secondHyper))
    }

    @Test
    fun `empty cells count as no hypercube and never crash`() {
        val hyper = gem(GemType.BLUE, Special.HYPERCUBE)
        assertFalse(bufferedSwapIsStale(null, null, null, null))
        assertTrue(bufferedSwapIsStale(null, null, hyper, null))
        assertTrue(bufferedSwapIsStale(hyper, null, null, null))
    }
}

File: Tests/BufferedSwapGuardTest.cs
using MatchThree.Engine.Model;
using MatchThree.Engine.Rules;
using NUnit.Framework;

namespace MatchThree.Engine.Tests;

public class BufferedSwapGuardTest
{
    private int nextId;

    private Gem Gem(GemType type = GemType.Red, Special? special = null) => new(nextId++, type, special);

    [Test]
    public void PlainPairWhoseGemsChangedEntirelyIsNotStale()
    {
        var submitA = Gem();
        var submitB = Gem();
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(submitA, submitB, Gem(), Gem()), Is.False);
    }

    [Test]
    public void HypercubeStillInThePairAtSettleIsNotStale()
    {
        var hyper = Gem(GemType.Blue, Special.Hypercube);
        var plain = Gem();
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(hyper, plain, hyper, plain), Is.False);
    }

    [Test]
    public void HypercubeBornOrFallenIntoThePairIsStale()
    {
        var plainA = Gem();
        var plainB = Gem();
        var settledHyper = Gem(GemType.Blue, Special.Hypercube);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(plainA, plainB, settledHyper, plainB), Is.True);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(plainA, plainB, plainA, settledHyper), Is.True);
    }

    [Test]
    public void HypercubeThatLeftThePairIsStale()
    {
        var hyper = Gem(GemType.Blue, Special.Hypercube);
        var plain = Gem();
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(hyper, plain, Gem(), Gem()), Is.True);
    }

    [Test]
    public void HypercubeFallingWithinThePairKeepsTheSwapFresh()
    {
        var hyper = Gem(GemType.Blue, Special.Hypercube);
        var plain = Gem();
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(hyper, plain, plain, hyper), Is.False);
    }

    [Test]
    public void ASecondHypercubeEnteringThePairIsStale()
    {
        var hyper = Gem(GemType.Blue, Special.Hypercube);
        var plain = Gem();
        var secondHyper = Gem(GemType.Green, Special.Hypercube);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(hyper, plain, hyper, secondHyper), Is.True);
    }

    [Test]
    public void EmptyCellsCountAsNoHypercubeAndNeverCrash()
    {
        var hyper = Gem(GemType.Blue, Special.Hypercube);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(null, null, null, null), Is.False);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(null, null, hyper, null), Is.True);
        Assert.That(BufferedSwapGuard.BufferedSwapIsStale(hyper, null, null, null), Is.True);
    }
}